larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,451 @@
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewThreeDKinkTool::TwoViewThreeDKinkTool() :
+    m_minXOverlapFraction(0.1f),
+    m_minMatchingScore(0.95f),
+    m_minLocallyMatchedFraction(0.3f),
+    m_minLongitudinalImpactParameter(-1.f),
+    m_nLayersForKinkSearch(10),
+    m_additionalXStepForKinkSearch(0.01f),
+    m_maxTransverseImpactParameter(5.f),
+    m_minImpactParameterCosTheta(0.5f),
+    m_cosThetaCutForKinkSearch(0.75f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TwoViewThreeDKinkTool::~TwoViewThreeDKinkTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::Run(TwoViewTransverseTracksAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    ModificationList modificationList;
+    this->GetModifications(pAlgorithm, overlapMatrix, modificationList);
+    const bool changesMade(this->ApplyChanges(pAlgorithm, modificationList));
+
+    return changesMade;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::PassesElementCuts(MatrixType::ElementList::const_iterator eIter, const ClusterSet &usedClusters) const
+{
+    if (usedClusters.count(eIter->GetCluster1()) || usedClusters.count(eIter->GetCluster2()))
+        return false;
+
+    if (eIter->GetOverlapResult().GetTwoViewXOverlap().GetXOverlapFraction0() - m_minXOverlapFraction < -std::numeric_limits<float>::epsilon())
+        return false;
+
+    if (eIter->GetOverlapResult().GetTwoViewXOverlap().GetXOverlapFraction1() - m_minXOverlapFraction < -std::numeric_limits<float>::epsilon())
+        return false;
+
+    if (eIter->GetOverlapResult().GetMatchingScore() - m_minMatchingScore < std::numeric_limits<float>::epsilon())
+        return false;
+
+    if (eIter->GetOverlapResult().GetLocallyMatchedFraction() - m_minLocallyMatchedFraction < std::numeric_limits<float>::epsilon())
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float TwoViewThreeDKinkTool::GetXSamplingPoint(const CartesianVector &splitPosition1, const bool isForwardInX,
+    const TwoDSlidingFitResult &fitResult1, const TwoDSlidingFitResult &fitResult2) const
+{
+    // Nearest common x position
+    float xMin1(std::numeric_limits<float>::max()), xMax1(-std::numeric_limits<float>::max());
+    float xMin2(std::numeric_limits<float>::max()), xMax2(-std::numeric_limits<float>::max());
```

Comment:
would `std::numeric_limits<float>::lowest()`, rather than `-std::numeric_limits<float>::max()` be appropriate here?

Reply:
I think you're right :)